interpreter/src/main/java/org/qbicc/interpreter/impl/Frame.java

```diff
@@ -314,8 +315,14 @@ public Object visit(VmThreadImpl thread, Add node) {
         ValueType inputType = node.getLeftInput().getType();
         assertSameTypes(node);
         if (isInt64(inputType)) {
-            // long math
-            return box(unboxLong(node.getLeftInput()) + unboxLong(node.getRightInput()), node.getType());
+            Object raw = require(node.getLeftInput());
+            if (raw instanceof MemoryPointer mp) {
+                // pointer arithmetic
+                return new MemoryPointer(mp.getType(), MemoryFactory.offset(mp.getMemory(), unboxLong(node.getRightInput())));
```

Comment:
I think this doesn't need to be special-cased for `MemoryPointer` - given `Pointer p` you should be able to do `p.offsetInBytes(offset, true)`.

Reply:
ok, that made this lot simpler of a patch :) 

Revised and squashed to a single commit since the digression of a specialized backing memory isn't needed.